### DISCUSSION TITLE
Route tree types

### DISCUSSION THREAD

**Discussion by tigawanna** [TanStack/router#824] [Q&A] [answered]
is there a way to get a type of all the possible routes 

```tsx

import { z } from 'zod'
import { Route, useRouter } from "@tanstack/react-router";
import { authlayout } from "./AuthLayout";
import { SignInForm } from "./components/SignInForm";
import { router } from '@/main';



const signinRouteSearchParamsSchema = z.object({
  redirect: z.string().optional(),
  email: z.string().email().optional(),
});

export const signinRoute = new Route({
  getParentRoute: () => authlayout,
  path: "/",
  validateSearch: (search) => {
    return signinRouteSearchParamsSchema.parse(search);
  },

  component: ({}) => {
    const router = useRouter();
    return (
      <div className="w-full h-full flex items-center justify-center">
        <SignInForm />
      </div>
    );
  },
});

```

in my case am trying to make sure you can only pass valid routes to 

```ts
  redirect: z.string().optional(),
```
instead of just a string

**Comment by schiller-manuel** [comment]
How would a _type_ of all possible routes help? Wouldn't you need that information at runtime?

**Reply by tigawanna** [reply]
Was creating a route helper function that generates breadcrumbs links , having all valid routes would help in typing that since I want it to sit at the root layout and dynamically grow and shrink based on the current route

**Reply by schiller-manuel** [reply, marked as answer]
How about something like this:

```typescript
type ValidRoutes = ParseRoute<typeof routeTree>['fullPath'];

// this results in for examples/react/kitchen-sink/src/main.tsx
type ValidRoutes = "/" | "/dashboard" | "/dashboard/" | "/dashboard/invoices" | "/dashboard/invoices/" | "/dashboard/invoices/$invoiceId" | "/dashboard/users" | "/dashboard/users/" | "/dashboard/users/user" | "/expensive" | "/profile" | "/login" | "/layout-a" | "/layout-b"
```

**Reply by tigawanna** [reply]
Yes I wanted something like this , thank you

**Reply by Michota** [reply]
> Was creating a route helper function that generates breadcrumbs links , having all valid routes would help in typing that since I want it to sit at the root layout and dynamically grow and shrink based on the current route

For me it says that `Type alias 'ValidPaths' circularly references itself.ts(2456)`

**Reply by arishem51** [reply]
> > Was creating a route helper function that generates breadcrumbs links , having all valid routes would help in typing that since I want it to sit at the root layout and dynamically grow and shrink based on the current route
> 
> For me it says that `Type alias 'ValidPaths' circularly references itself.ts(2456)`

same for me, any solution pls?

**Reply by schiller-manuel** [reply]
can you please provide a minimal complete example by forking one of the existing router examples on stackblitz?

**Reply by nzyoni** [reply]
> > Was creating a route helper function that generates breadcrumbs links , having all valid routes would help in typing that since I want it to sit at the root layout and dynamically grow and shrink based on the current route
> 
> For me it says that `Type alias 'ValidPaths' circularly references itself.ts(2456)`

same here 😢

**Reply by yoshintame** [reply]
same

**Reply by schiller-manuel** [reply]
without any reproducers we won't be able to help you...

**Reply by JosueGalRe** [reply]
I'm also facing this issue, it is weird, it was working perfectly a few days ago, and now it doesn't
<img width="862" alt="image" src="https://github.com/user-attachments/assets/db94e573-be23-4d13-a3a8-6fffefc59780">

on stackblitz:
<img width="676" alt="image" src="https://github.com/user-attachments/assets/3d1f755b-bc17-4432-bab0-ad79867111c3">

i couldn't reproduce it, tried matching deps versions and ts config and nothing, maybe someone can try
https://stackblitz.com/edit/tanstack-router-aldgoz?file=src%2Ftypes%2Futils.ts

**Reply by JosueGalRe** [reply]
Never mind, this is a working issue reproduction: https://stackblitz.com/~/github.com/JosueGalRe/tanstack-router-issue-repro @schiller-manuel

**Reply by schiller-manuel** [reply]
@JosueGalRe you have a circular type reference.

`schema` in `-index.utils.tsx` references `ValidRoutes` which is built by parsing the routes which again needs to look into `schema` by checking the type of `validateSearch` in `index.ts`.

for your particular use case, you could use this:

```tsx
import * as v from 'valibot';

import type { FileRoutesByFullPath } from 'routeTree.gen';

export const schema = v.object({
  redirectTo: v.optional(
    v.custom<keyof FileRoutesByFullPath>((value) => {
      return true;
    }),
    '/'
  ),
});

export type Schema = v.InferInput<typeof schema>;
```

please be aware that this is not performing any validation at runtime

**Reply by JosueGalRe** [reply]
@schiller-manuel thanks, I didn't think about this, but is there any way I can validate the routes runtime? Like an util to retrieve all routeTree routes in a single array?

**Reply by schiller-manuel** [reply]
you could check `router.routesByPath`, but accessing that in the validator would again introduce a circular type reference :(

so you need to trick the compiler:

```tsx
import * as v from 'valibot';

import type { FileRoutesByFullPath } from 'routeTree.gen';

// need to export router from main.tsx!
import { router } from 'main';


export const schema = v.object({
  redirectTo: v.optional(
    v.custom<keyof FileRoutesByFullPath>((value) => {
      // without `as any` we would again get circular type reference errors
      return (router as any).routesByPath[value as any] !== undefined
    }),
    '/'
  ),
});

export type Schema = v.InferInput<typeof schema>;
```

**Reply by schiller-manuel** [reply]
BTW, your route tree configuration is not correct. you have both a `/index.tsx` and a `/_auth/index.tsx`.
Both would lead to the `/` path at runtime.

The router generator did not check this, now it does and will throw an error: https://github.com/TanStack/router/pull/2525

**Reply by JosueGalRe** [reply]
@schiller-manuel Thank you so much!

**Comment by muneer-abdulsattar** [comment]
i used to use this and works fine

```js
import {ParseRoute} from "@tanstack/react-router";`
`export type TRoutes = ParseRoute<typeof routeTree>["fullPath"];
```


but now i keep getting error of 

```ts
Module '"@tanstack/react-router"' has no exported member 'ParseRoute'.ts(2305)
```

**Reply by muneer-abdulsattar** [reply]
i manage to get it to work using

```ts
import { type FileRoutesByTo } from "@/routeTree.gen";
export type TRoutes = keyof FileRoutesByTo;
```

**Reply by Atchferox** [reply]
This is the real correct answer :)

**Reply by tigawanna** [reply]
can confirm as 
```ts
import { ParseRoute } from "@tanstack/react-router";
```
was remoevd as of the latest version

**Reply by schiller-manuel** [reply]
those types are in router-core now.
however, they are still internal types only.
we have public types here: https://tanstack.com/router/latest/docs/framework/react/guide/type-utilities
does this help?

**Reply by shlroland** [reply]
> i manage to get it to work using
> 
> ```ts
> import { type FileRoutesByTo } from "@/routeTree.gen";
> export type TRoutes = keyof FileRoutesByTo;
> ```

Are there any good methods for code-based routing?

**Reply by igorkoho** [reply]
@schiller-manuel are there any plans to make a public type of all the valid route paths? It is quite useful to be able to make sure some strings in business logic consist of only valid route paths, for example when doing route guards that want to have specific logic based on which sub-route the user is on

**Comment by GrahamQuan** [comment]
i use version `"@tanstack/react-router": "^1.130.2"`

```ts
import type { FileRouteTypes } from '@/routeTree.gen'

export type NavItem = {
  name: string
  href: FileRouteTypes['to'] // or FileRouteTypes['fullPaths']
}
```

i think the source code already do `keyof` for us

**Reply by clemwo** [reply]
Is this supposed to work with `getRouteApi` as well?

I have the following react hook code that still has an error:

```
import {
  type AnyRouter,
  getRouteApi,
  type LinkProps,
  type RegisteredRouter,
} from '@tanstack/react-router';

type Routes<TRouter extends AnyRouter = RegisteredRouter> =
  LinkProps<TRouter>['to'];

const useRouteById = (routeId: Routes) => {
  const route = getRouteApi(routeId);
  //   do stuff with route here
  return;
};

```  


Anyone has an idea on how to type the routeId correctly?

**Reply by tigawanna** [reply]
For that use this 
```ts
import type { FileRoutesById } from "@/routeTree.gen";
type RouteIDsType = keyof FileRoutesById;
```

**Comment by dheardal** [comment]
A more generic way to get the list of navigation routes, for when you don't have access to the route gen file:

```
import { AnyRouter, LinkProps, RegisteredRouter } from '@tanstack/react-router';
type Routes<TRouter extends AnyRouter = RegisteredRouter> = LinkProps<TRouter>['to'];
```

**Comment by sergey-lipikhin** [comment]
```TypeScript
import type { FileRouteTypes } from '@/routeTree.gen';

type RoutePath = FileRouteTypes['fullPaths'];
```
